cores/arduino/HardwareTimer.cpp

```diff
@@ -0,0 +1,1429 @@
+  * @param  htim : one of the defined timer
+  * @retval None
+  */
+timer_index_t get_timer_index(TIM_TypeDef *instance)
+{
+  timer_index_t index = UNKNOWN_TIMER;
+
+#if defined(TIM1_BASE)
+  if (instance == TIM1) {
+    index = TIMER1_INDEX;
+  }
+#endif
+#if defined(TIM2_BASE)
+  if (instance == TIM2) {
+    index = TIMER2_INDEX;
+  }
+#endif
+#if defined(TIM3_BASE)
+  if (instance == TIM3) {
+    index = TIMER3_INDEX;
+  }
+#endif
+#if defined(TIM4_BASE)
+  if (instance == TIM4) {
+    index = TIMER4_INDEX;
+  }
+#endif
+#if defined(TIM5_BASE)
+  if (instance == TIM5) {
+    index = TIMER5_INDEX;
+  }
+#endif
+#if defined(TIM6_BASE)
+  if (instance == TIM6) {
+    index = TIMER6_INDEX;
+  }
+#endif
+#if defined(TIM7_BASE)
+  if (instance == TIM7) {
+    index = TIMER7_INDEX;
+  }
+#endif
+#if defined(TIM8_BASE)
+  if (instance == TIM8) {
+    index = TIMER8_INDEX;
+  }
+#endif
+#if defined(TIM9_BASE)
+  if (instance == TIM9) {
+    index = TIMER9_INDEX;
+  }
+#endif
+#if defined(TIM10_BASE)
+  if (instance == TIM10) {
+    index = TIMER10_INDEX;
+  }
+#endif
+#if defined(TIM11_BASE)
+  if (instance == TIM11) {
+    index = TIMER11_INDEX;
+  }
+#endif
+#if defined(TIM12_BASE)
+  if (instance == TIM12) {
+    index = TIMER12_INDEX;
+  }
+#endif
+#if defined(TIM13_BASE)
+  if (instance == TIM13) {
+    index = TIMER13_INDEX;
+  }
+#endif
+#if defined(TIM14_BASE)
+  if (instance == TIM14) {
+    index = TIMER14_INDEX;
+  }
+#endif
+#if defined(TIM15_BASE)
+  if (instance == TIM15) {
+    index = TIMER15_INDEX;
+  }
+#endif
+#if defined(TIM16_BASE)
+  if (instance == TIM16) {
+    index = TIMER16_INDEX;
+  }
+#endif
+#if defined(TIM17_BASE)
+  if (instance == TIM17) {
+    index = TIMER17_INDEX;
+  }
+#endif
+#if defined(TIM18_BASE)
+  if (instance == TIM18) {
+    index = TIMER18_INDEX;
+  }
+#endif
+#if defined(TIM19_BASE)
+  if (instance == TIM19) {
+    index = TIMER19_INDEX;
+  }
+#endif
+#if defined(TIM20_BASE)
+  if (instance == TIM20) {
+    index = TIMER20_INDEX;
+  }
+#endif
+#if defined(TIM21_BASE)
+  if (instance == TIM21) {
+    index = TIMER21_INDEX;
+  }
+#endif
+#if defined(TIM22_BASE)
+  if (instance == TIM22) {
+    index = TIMER2_INDEX;
+  }
+#endif
+  return index;
+}
+
+/**
+  * @brief  This function return the timer clock frequency.
+  * @param  tim: timer instance
+  * @retval frequency in Hz
+  */
+uint32_t HardwareTimer::getTimerClkFreq()
+{
+  RCC_ClkInitTypeDef    clkconfig = {};
+  uint32_t              pFLatency = 0U;
+  uint32_t              uwTimclock = 0U, uwAPBxPrescaler = 0U;
+
+  /* Get clock configuration */
+  HAL_RCC_GetClockConfig(&clkconfig, &pFLatency);
+  switch (getTimerClkSrc(_HardwareTimerObj.handle.Instance)) {
+    case 1:
+      uwAPBxPrescaler = clkconfig.APB1CLKDivider;
+      uwTimclock = HAL_RCC_GetPCLK1Freq();
+      break;
+#if !defined(STM32F0xx) && !defined(STM32G0xx)
+    case 2:
+      uwAPBxPrescaler = clkconfig.APB2CLKDivider;
+      uwTimclock = HAL_RCC_GetPCLK2Freq();
+      break;
```

Comment:
Hey @ABOSTM, do you remember why you added this exception (that PCLK2 is not supported on F0 and G0)? I just ran into TIM1 not working on a G030, and found this bit of code. It does seem that F0 and G0 actually have PCLK2 (I checked F030, G0x0 and G0x1 reference manuals). Is there maybe some code missing to enable this clock or so? Maybe @fpistm knows?

Reply:
Hi @matthijskooijman 
`HAL_RCC_GetPCLK2Freq` does not exist for those series.
I've checked the reference manual for STM32G0x0 and found no PCLK2 only TIMPCLK.
![image](https://user-images.githubusercontent.com/20641798/125754156-64926947-1549-4c63-8e16-a347e58ce7f8.png)

Reply:
@matthijskooijman , this is an inheritance of `getTimerClkFreq()` from timer.c before implementation of HardwareTimer.
but this doesn't help you much :-)
The real reason behind is that this `case 2` concerns TIMERs that are on APB2 bus, but looking at STM32F030 there is no ABP2. so the function `HAL_RCC_GetPCLK2Freq()` doesn't exist on G0 and F0

Reply:
NVM, we seem to have hit an older bug from a previous core version where `getTimerClkSrc()` returned a clock source 2 instead of 1 and since case 2 is purposely not defined for the F0/G0 in `getTimerClkFreq()` it would end up in case 0's `Error_handler()`  :-)
This has been fixed in the commit [here](https://github.com/stm32duino/Arduino_Core_STM32/commit/0a0b2b98389bdff440d3a9d1ac6a2b0ad07f28dd) (core version 2.0.0).